packages/core/src/anchor/anchor-processing-loop.ts

```diff
@@ -0,0 +1,60 @@
+import { ProcessingLoop } from './processing-loop.js'
+import type { CASClient } from './anchor-service.js'
+import type {
+  AnchorRequestStore,
+  AnchorRequestStoreListResult,
+} from '../store/anchor-request-store.js'
+import { AnchorRequestCarFileReader } from './anchor-request-car-file-reader.js'
+import type { AnchorLoopHandler } from './anchor-service.js'
+import type { DiagnosticsLogger } from '@ceramicnetwork/common'
+
+/**
+ * Get anchor request entries from AnchorRequestStore one by one. For each entry, get CAS response,
+ * and handle the response via `eventHandler.handle`.
+ *
+ * If CAS does not know about an anchor request, we consider it being not created previously for some reason.
+ * We create a request on CAS then.
+ */
+export class AnchorProcessingLoop {
+  #loop: ProcessingLoop<AnchorRequestStoreListResult>
+  constructor(
+    batchSize: number,
+    cas: CASClient,
+    store: AnchorRequestStore,
+    logger: DiagnosticsLogger,
+    eventHandler: AnchorLoopHandler
+  ) {
+    this.#loop = new ProcessingLoop(store.infiniteList(batchSize), async (entry) => {
+      try {
+        const event = await cas.getStatusForRequest(entry.key, entry.value.cid).catch(async () => {
+          const requestCAR = await eventHandler.buildRequestCar(entry.key, entry.value.cid)
+          return cas.create(new AnchorRequestCarFileReader(requestCAR), false)
+        })
+        const isTerminal = await eventHandler.handle(event)
+        if (isTerminal) {
+          // We might store a new entry during the processing, so we better check if the current entry is indeed terminal.
+          const current = await store.load(entry.key)
+          if (current.cid.equals(entry.value.cid)) {
```

Comment:
won't this check be racy?  What if there is a write to the request store in flight already at this point?  I think all reads/writes to the anchor request store for a stream need to be serialized in the execution queue to be safe

Reply:
Maybe if we can push this down into the eventHandler itself and do it all within the context of the execution queue?